### PR TITLE
Merge folders within the explorer rather than overwrite

### LINKED ISSUE #165706: Moving a folder to a location with it already existing deletes the existing foler
Type: <b>Bug</b>

Create the Following File system: 
Folder1
 | - Folder2
 |   | - File1
Folder2
 | - File2

Now move the lower `Folder2` into `Folder1`. (via Drag-n-Drop)
Instead of the contents of both Folders being merged only the contents of the lower folder are there after the move operation. 

VS Code version: Code 1.72.2 (d045a5eda657f4d7b676dedbfa7aab8207f8a075, 2022-10-12T22:15:18.074Z)
OS version: Windows_NT x64 10.0.22621
Modes:
Sandboxed: No
Remote OS version: Linux x64 5.4.0-1094-azure

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|undefined|
|Memory (System)|15.92GB (3.28GB free)|
|Process Argv|--crash-reporter-id c0b728c1-554c-4067-b8ee-a399c07e544b|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|Codespaces|
|OS|Linux x64 5.4.0-1094-azure|
|CPUs|Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz (4 x 2593)|
|Memory (System)|7.77GB (3.77GB free)|
|VM|0%|
</details><details><summary>Extensions (64)</summary>

Extension|Author (truncated)|Version
---|---|---
packer|4op|0.3.0
auto-rename-tag|for|0.1.10
codespaces|Git|1.13.1
gitpod-desktop|git|0.0.64
jupyter-keymap|ms-|1.0.0
remote-containers|ms-|0.262.3
remote-ssh|ms-|0.90.1
remote-ssh-edit|ms-|0.84.0
remote-wsl|ms-|0.72.0
vscode-remote-extensionpack|ms-|0.21.0
riscv-plus|ste|0.1.0
vscode-icons|vsc|12.0.1
riscv|zhw|0.0.8
vscode-openapi|42C|4.15.9
packer|4op|0.3.0
vscode-base64|ada|0.1.0
ng-template|Ang|14.2.0
vscode-eslint|dba|2.2.6
xml|Dot|2.5.1
gitlens|eam|13.0.4
EditorConfig|Edi|0.16.4
prettier-vscode|esb|9.9.0
auto-rename-tag|for|0.1.10
codespaces|Git|1.13.1
copilot|Git|1.57.7193
vscode-pull-request-github|Git|0.50.0
gitpod-remote-ssh|git|0.0.38
go|gol|0.35.2
todo-tree|Gru|0.0.219
rest-client|hum|0.25.1
latex-workshop|Jam|9.0.0
docomment|k--|0.1.31
rainbow-csv|mec|3.3.0
file-downloader|min|1.0.12
vscode-docker|ms-|1.22.2
csharp|ms-|1.25.0
python|ms-|2022.16.1
vscode-pylance|ms-|2022.11.10
jupyter|ms-|2022.9.1202862440
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.10
vscode-jupyter-cell-tags|ms-|0.1.6
vscode-jupyter-powertoys|ms-|0.0.6
vscode-jupyter-slideshow|ms-|0.1.5
azure-account|ms-|0.11.2
cpptools|ms-|1.12.4
hexeditor|ms-|1.9.9
live-server|ms-|0.4.3
vscode-typescript-tslint-plugin|ms-|1.3.4
vsliveshare|ms-|1.0.5762
quicktype|qui|12.0.46
java|red|1.12.0
vscode-commons|red|0.0.6
vscode-yaml|red|1.10.1
riscv-plus|ste|0.1.0
wolf|tra|0.4.3
vscodeintellicode|Vis|1.2.29
vscode-java-debug|vsc|0.46.0
vscode-java-dependency|vsc|0.21.1
vscode-java-pack|vsc|0.25.6
vscode-java-test|vsc|0.37.1
vscode-icons|vsc|12.0.1
quokka-vscode|Wal|1.0.508
riscv|zhw|0.0.8

(2 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242:30382549
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
vscoreces:30445986
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593:30376534
pythonvs932:30410667
cppdebug:30492333
vsclangdf:30486550
c4g48928:30535728
dsvsc012cf:30540253
azure-dev_surveyone:30548225
pyindex848:30577860
nodejswelcome1:30587005
fc301958:30595537
282f8724:30602487
iaj6b796:30604794
dbltrim-noruby:30604474

```

</details>

### PR DESCRIPTION
**Pull request by lramos15**
Fixes #165706

### CODE REVIEW

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts [hunk 1]
```diff
@@ -1362,6 +1363,23 @@ export class FileDragAndDrop implements ITreeDragAndDrop<ExplorerItem> {
...
 	}
 
+	private async mergeDirectories(sources: ExplorerItem[], targetResource: URI): Promise<ResourceFileEdit[]> {
+		const resourceFileEdits: ResourceFileEdit[] = [];
+		for (const source of sources) {
+			const newResource = joinPath(targetResource, source.name);
+			// If the directory matches then we should recurse on the children because we want to merge them
+			if (source.isDirectory && await this.fileService.exists(newResource)) {
+				// Recurse into the children of the directories so that we can merge them with the source
+				resourceFileEdits.push(...await this.mergeDirectories(Array.from(source.children.values()), newResource));
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1373]: Will `source.children` work even when the folder is not expanded in the Explorer? My understanding would be that we would have to also resolve explorer items recursively to ensure we have all the children recursively?

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts [hunk 2]
```diff
@@ -1325,7 +1325,8 @@ export class FileDragAndDrop implements ITreeDragAndDrop<ExplorerItem> {
 	private async doHandleExplorerDropOnMove(sources: ExplorerItem[], target: ExplorerItem): Promise<void> {
 
 		// Do not allow moving readonly items
-		const resourceFileEdits = sources.filter(source => !source.isReadonly).map(source => new ResourceFileEdit(source.resource, joinPath(target.resource, source.name)));
+		sources = sources.filter(s => !s.isReadonly);
+		const resourceFileEdits = await this.mergeDirectories(sources, target.resource);
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1329]: Should this be something the user is asked whether to merge or to replace?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts (hunk 2 above)
Comment (lramos15) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1329]: We can ask as that is one of the proposals in this issue. I feel like merging is pretty risk free, but it's not the default behavior on all OSes. On macOS it's only the default if you hold option whereas Windows it's always the default.

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts (hunk 1 above)
Comment (lramos15) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts]: Good catch, I now use the async method `fetchChildren` which should actually resolve them in the model.

**Comment by lramos15** [comment]
@bpasero I think this is ready for a re-review. Let me know if you think we should take this in for this iteration or wait until the next 👍

**Comment by bpasero** [comment]
Well, there is commented out lines and `console.log` statements in the change, so I had assumed this is still in draft mode.

**Comment by lramos15** [comment]
> Well, there is commented out lines and `console.log` statements in the change, so I had assumed this is still in draft mode.

Oops. That was my bad for not checking those were gone, this has been fixed now. Thanks for pointing them out.

**Review by bpasero — CHANGES_REQUESTED** [review]
There seems to be inconsistent capital casing in the buttons:

image

Also, I only was able to get this dialog when doing a DND operation inside the explorer, but why does this not work when I drop a folder from macOS finder into the explorer?

**Comment by vivodi** [comment]
Is this still relevant?

**Comment by uditkotecha1-hub** [comment]
🕵️ CodeSherlock Review:

🔴 Critical Issues:
- No critical issues found.

🟡 Improvements:
- Consider adding error handling for operations like merging directories or applying bulk edits. This can help inform the user if something goes wrong.

🟢 Good Practices:
- The new `MergePromptChoice` enum improves code readability and maintainability by providing meaningful names for the prompt choices.
- The usage of `async/await` for asynchronous operations enhances code clarity.

💡 Suggestions:
- It would be beneficial to add documentation comments to the newly added methods (`mergeDirectories` and `promptForMergeOrOverwrite`) to describe their functionality.
- Consider unit tests for the new merging logic to ensure it handles various edge cases effectively, such as source directories being empty or already existing in the target.

Risk Level: MEDIUM
